lib/AST/ASTContext.cpp

```diff
@@ -7739,6 +7740,11 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
       if (paramTy->isPromotableIntegerType() ||
           getCanonicalType(paramTy).getUnqualifiedType() == FloatTy)
         return QualType();
+
+      // For Checked C, a no prototype function is not compatible
+      // with a prototype with a checked argument.
+      if (isCheckedC && isNotAllowedForNoProtoTypeFunction(paramTy))
```

Comment:
There's an if-statement at line 7566 that handles the case where `lproto` and `rproto` are both non-null.    The code under `if (proto)` handles the case where one of them is null (has no prototype) and the other is non-null (has a prototype).